docs/rfc/0698-cli-infrastructure.adoc

```diff
@@ -0,0 +1,295 @@
+exhaustive list:
+
+* `global-verbose`
+* `global-quiet`
+* `global-profile := active (default) | <profile-id>`
+
+== CLI Events
+
+The execution of a CLI subcommand could potentially result in an
+interesting event. For example, if Upstream is
+running and I create a project via the CLI, then a good user
+experience is that the Upstream is updated with the new project
+(without refreshing). In this case, the event is the creation of the
+project.
+
+In <<arch-pubsub, RFC-682#PubSub>>, a PubSub component was proposed --
+with a detailed specification being left to a future RFC. We will
+assume the existence of this component here. The CLI commands in this
+RFC MAY publish interesting events to the PubSub component. This will
+allow other applications to react upon seeing these events via
+subscriptions to the PubSub component.
+
+Other subcommands MAY also publish interesting events, as long as they
+adhere to the specification of the future RFC. They SHALL reuse any
+events defined as part of the <<Core>> commands. That is to say, there
+will be a core set of events that should not be overridden, but rather
+reused if needs be.
+
+== Key Access
+
+<<arch-sigs, RFC-682#Signatures>> outlined the signing of Radicle data. This is essential to
+this RFC as commands that write to the storage generally need access
+to the secret key, whether for signing data or proof of ownership to
+the storage. For most (if not all) read-only commands, no access to
+the secret key is needed.
+
+To achieve the goals of having a read-only storage and using
+`<<ssh-agent>>` for signing, we propose the following should be
+implemented:
+
+. `ssh-agent` compatability with `<<radicle-keystore>>`, which has been
+started in `<<rk-17, radicle-keystore#17>>`.
+. A subcommand `rad-profile-ssh-add` for adding one's secret key to the
+`ssh-agent` using the supplied constraints, e.g. ask for password.
+. Implement `<<rl-461, radicle-link#461>>` to have a read-only storage.
+
+== Library/Command Organisation
+
+As proposed here, `radicle-link` will provide a core set of
+functionality. Something this document wishes to standardise is the
+organisation of these components. The author sees two options:
+
+. One binary that defines the core subcommands
+. Many binaries that take advantage of the subcommand mechanism
+
+Pros of *1.*:
+
+* One-stop shop for all core functionality
+* One binary is packaged and distributed
+
+Cons of *1.*:
+
+* Making changes to one subset of the commands could mean a version
+  change of the whole binary
+
+Pros of *2.*:
+
+* We will be organising commands as libraries (as per RFC-682) — so
+  modular commands would mirror this structure well.
+* Subcommands can be versioned independently of each other.
+* We can dogfood our subcommand functionality
+
+Cons of *2.*:
+
+* Makes it easier to override one of the core commands.
+* Many binaries must be packaged and distributed.
+
+The author recommends *1.*. As mentioned, we have a hard requirement
```

Comment:
There is also a hybrid approach, exemplified by `git`: the “builtin” executables are deployed to `GIT_EXEC_PATH`, which is by default not a user-writable directory. Security-wise, this is only marginally better than `PATH`, but perhaps if this path could only be overridden by a `—exec-path` argument it would be acceptable.

I still don’t, for the set of commands described here, see a compelling reason why release builds should be split into many binaries, but I do see how this could be handy during development.

I would say the simple thing is a single binary, and user commands a la cargo.

Reply:
Ya, I'm happy with what we landed on. Leaving the prose here for documentation of at least considering it.